### PR TITLE
Added OpInfo-based testing of some linalg functions

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Added OpInfo-based testing of the following linear algebra functions:
* cholesky, linalg.cholesky
* linalg.eigh
* inverse, linalg.inv
* qr, linalg.qr
* solve

The output of `torch.linalg.pinv` for empty inputs was not differentiable, now it's fixed.

In some cases, batched grad checks are disabled because it doesn't work well with 0x0 matrices (see https://github.com/pytorch/pytorch/issues/50743#issuecomment-767376085).

Ref. #50006

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
Tests for QR decomposition are quite slow
```
========================================================= slowest 20 durations =========================================================
32.76s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_qr_cuda_complex128
32.08s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_qr_cuda_complex128
16.98s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_qr_cuda_float64
15.77s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_solve_cuda_complex64
15.09s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_solve_cuda_complex128
14.89s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_qr_cuda_float32
13.98s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_solve_cuda_float32
13.53s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_flip_cuda_complex128
13.19s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_qr_cuda_complex64
13.05s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_svd_cuda_complex128
13.04s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_svd_cuda_complex128
12.51s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_svd_cuda_float64
11.87s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_qr_cuda_complex128
10.38s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_solve_cuda_float32
10.06s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_solve_cuda_complex64
9.71s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_solve_cuda_complex128
9.08s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_qr_cuda_float64
9.05s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_qr_cuda_float64
8.80s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_qr_cuda_float64
8.68s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_svd_cuda_float32
===================================== 5139 passed, 1009 skipped, 61 warnings in 1069.53s (0:17:49) =====================================
```

**Comment by IvanYashchuk** [comment]
The failing tests should be fixed after merging https://github.com/pytorch/pytorch/pull/51109.

**Comment by mruberry** [comment]
Hey @IvanYashchuk! Just checking-in on this PR. Is it waiting for review?

**Comment by IvanYashchuk** [comment]
> Hey @IvanYashchuk! Just checking-in on this PR. Is it waiting for review?

It's ready for review. I was just not pinging you knowing you're busy and hoping that this PR is in your queue :wink:

**Comment by mruberry** [comment]
> > Hey @IvanYashchuk! Just checking-in on this PR. Is it waiting for review?
> 
> It's ready for review. I was just not pinging you knowing you're busy and hoping that this PR is in your queue 😉

Haha, in my queue enough to ask you about it. Taking a look now.

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1909]: Is there a better name for sample_inputs_linalg_inv if it's being used for slogdet and pinv, too?

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1550]: This is fine although it'd be nice if HermitianOpInfo applied this skip so each HermitianOpInfo didn't need to.

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2180]: I appreciate this TODO but this is more a failure of our out= testing (which we should be updating soonish) and not an issue with torch.solve().

**Review by mruberry — APPROVED** [review]
Awesome! Nice work, @IvanYashchuk. I really appreciate how thorough and consistent you were with documentation. It makes the code much more readable and easier to maintain.

This needs a rebase. Just ping me when it's ready to merge.

File: torch/testing/_internal/common_methods_invocations.py
Comment (IvanYashchuk) [reply torch/testing/_internal/common_methods_invocations.py]: I agree, I moved this skip to HermitianOpInfo and applied a similar change for TriangularOpInfo.

File: torch/testing/_internal/common_methods_invocations.py
Comment (IvanYashchuk) [reply torch/testing/_internal/common_methods_invocations.py]: I changed it to sample_inputs_linalg_invertible.

**Comment by IvanYashchuk** [comment]
@mruberry I updated and rebased this PR. It's ready to merge.

**Comment by mruberry** [comment]
The ROCm test failures on this PR are interesting and linalg related. Would you take a look at them before we merge this?

**Comment by IvanYashchuk** [comment]
It was my fault not being careful enough when resolving the merge conflict and I accidentally removed `skipCUDAIfRocm` for `logdet`. Let's wait for CI now.

**Comment by IvanYashchuk** [comment]
ROCm still fails for 
```
test_variant_consistency_jit_linalg_pinv_cuda_float32
test_variant_consistency_eager_triangular_solve_cuda_float32
test_variant_consistency_eager_linalg_pinv_hermitian_cuda_float32
test_variant_consistency_eager_linalg_pinv_cuda_float32
test_out_linalg_pinv_hermitian_cuda_float32
test_out_linalg_pinv_cuda_float32
```
```
RuntimeError: magma: The value of work_size(-9223372036854775808) is too large to fit into a magma_int_t (4 bytes)

----------------------------------------------------------------------
Ran 9205 tests in 1586.823s

FAILED (errors=6, skipped=2448)
```

**Comment by IvanYashchuk** [comment]
@mruberry I updated this pull request with the recent changes for "out" testing. I also added all relevant ROCm skips, which will be fixed sometime later. Could you take a look once again and hopefully merge this?

**Comment by mruberry** [comment]
I'll try to get this landed ASAP. I was concerned about the ROCm failure because it's a timeout, but that test build doesn't appear to be running these tests. The overall test time seems similar to current CI timings.

**Pull request closed without merging**
